File: 124.binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    private int res = Integer.MIN_VALUE;
    public int maxPathSum(TreeNode root) {
        getGain(root);
        return res;
    }
    private int getGain(TreeNode node) {
        if (node == null) {
            return 0;
        }
        int leftGain = getGain(node.left);
        int rightGain = getGain(node.right);
        this.res = Math.max(this.res, node.val+leftGain+rightGain);
        int gain = node.val + Math.max(leftGain, rightGain);
        return gain > 0 ? gain : 0;
    }
}

File: 124.binary-tree-maximum-path-sum.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */import "math"

func maxPathSum(root *TreeNode) int {
	res := math.MinInt32
	dfs(root, &res)
	return res
}

func dfs(node *TreeNode, res *int) int {
	if node == nil {
		return 0
	}
	value := node.Val
	leftGain := max(0, dfs(node.Left, res))
	value += leftGain
	rightGain := max(0, dfs(node.Right, res))
	value += rightGain
	*res = max(*res, value)
	return node.Val + max(leftGain, rightGain)
}

func max(i, j int) int {
	if i > j {
		return i
	}
	return j
}
